{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fea9653b-7cf2-41c5-81b9-3b239b0f6d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install kSpider2 kProcessor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e592bc86-315b-4b16-9ed3-33a4ad1b2a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "KSIZE=25\n",
    "\n",
    "VER38_GRCh38_URL=http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/gencode.v38.transcripts.fa.gz\n",
    "VER28_GRCh37_URL=http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_28/GRCh37_mapping/gencode.v28lift37.transcripts.fa.gz\n",
    "\n",
    "VER38_GRCh38=VER38_GRCh38.fa\n",
    "VER28_GRCh37=VER28_GRCh37.fa\n",
    "\n",
    "wget --quiet ${VER38_GRCh38_URL} -O ${VER38_GRCh38}.gz\n",
    "wget --quiet ${VER28_GRCh37_URL} -O ${VER28_GRCh37}.gz\n",
    "\n",
    "gunzip *gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607af8e6-4881-4f60-8ad0-6a0f8ca4ddd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f89ccd84-3d88-43f5-88af-0abc1c859152",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "VER38_GRCh38=VER38_GRCh38.fa\n",
    "VER28_GRCh37=VER28_GRCh37.fa\n",
    "\n",
    "#1 Transform first annotation -------------------------------------------------------------\n",
    "sed -i 's/^>/>VER38_GRCh38|/' ${VER38_GRCh38}\n",
    "\n",
    "grep \">\" ${VER38_GRCh38} | cut -c2- |  awk -F'|' '{print $0\"\\tver38_GRCh38.\"$3}' > ${VER38_GRCh38}.names\n",
    "\n",
    "#2 Transform second annotation\n",
    "sed -i 's/^>/>VER28_GRCh37|/' ${VER28_GRCh37}\n",
    "\n",
    "grep \">\" ${VER28_GRCh37} | cut -c2- |  awk -F'|' '{print $0\"\\tver28_GRCh37.\"$3}' > ${VER28_GRCh37}.names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a358888-0489-4267-bf18-417f68b56643",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d71e81-2ad1-4c1a-b6ab-646c347abfbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "VER38_GRCh38=VER38_GRCh38.fa\n",
    "VER28_GRCh37=VER28_GRCh37.fa\n",
    "\n",
    "cat ${VER38_GRCh38} ${VER28_GRCh37} > merged.fa\n",
    "cat ${VER38_GRCh38}.names ${VER28_GRCh37}.names > merged.fa.names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e435e7-92da-434e-addc-a6bc0366f7ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import kProcessor as kp\n",
    "\n",
    "kSize = 25\n",
    "chunkSize = 5000\n",
    "fasta_file = \"merged.fa\"\n",
    "\n",
    "KF = kp.kDataFramePHMAP(kp.KMERS, kp.nonCanonicalInteger_Hasher, {\"kSize\":kSize})\n",
    "cKF = kp.index(KF, fasta_file, chunkSize, fasta_file + \".names\")\n",
    "cKF.save(\"idx_\" + fasta_file.replace(\".fa\",\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eece265-6b86-44ef-9c2d-c55265433c50",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3913dd5d-2070-4845-9485-34e7b4e0e467",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "kSpider2 pairwise -i idx_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8735de7-48d7-49f2-970f-943f4959a89f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe5a576-f46e-4a92-85dc-3ebc5807df01",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "kPro_index=\"idx_merged\"\n",
    "paste <(tail -n+2 ${kPro_index}.namesMap |cut -d\" \" -f1)  <(tail -n+2 ${kPro_index}.namesMap |cut -d\" \" -f2-) > ${kPro_index}.namesMap.tmp\n",
    "echo \"node_id node_name size\" | tr ' ' '\\t' > ${kPro_index}_nodes_size.tsv\n",
    "awk 'BEGIN{FS=OFS=\"\\t\";}FNR==NR{a[$2]=$3;next;}{if(a[$1]!=\"\")print $0,a[$1]}' ${kPro_index}_kSpider_seqToKmersNo.tsv ${kPro_index}.namesMap.tmp >> ${kPro_index}_nodes_size.tsv\n",
    "rm ${kPro_index}.namesMap.tmp*\n",
    "\n",
    "# 2. Annotation of the numerically coded association file\n",
    "# a) Add items names and no of associated items \n",
    "# b) calc jaccard distance and containment ratio for each pair\n",
    "# c) Additionally, we can filter out those with minimal similarities \n",
    "md=20   ## minimum jaccard distance (as a percentage) to keep the record \n",
    "mc=80  ## minimum containment ratio (as a percentage) to keep the record \n",
    "echo \":START_ID|START_name|START_size|shared_count:int|jDist:float|smPerc:float|END_name|END_size|:END_ID\" > ${kPro_index}_relations.csv\n",
    "awk -v md=$md -v mc=$mc 'BEGIN{FS=\"\\t\";S=\"|\";}FNR==NR{a[$1]=$3;b[$1]=$2S$3;next;}{\n",
    "   g1=a[$2]; g2=a[$3]; min=g1;min=(min < g2 ? min : g2); \n",
    "   jDist=$4*100/(g1+g2-$4); smPerc=$4*100/min; \n",
    "   if(jDist>md || smPerc>mc)\n",
    "     printf(\"%s%s%s%s%s%s%.1f%s%.1f%s%s%s%s\\n\", $2,S,b[$2],S,$4,S,jDist,S,smPerc,S,b[$3],S,$3)}' \\\n",
    "   ${kPro_index}_nodes_size.tsv <(tail -n+2 ${kPro_index}_kSpider_pairwise.tsv) >> ${kPro_index}_relations.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9adca17e-a1ae-4de0-a184-e6600aa5086d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "856a832d-3458-416c-885c-ff3e2775e9af",
   "metadata": {},
   "outputs": [],
   "source": [
    "tr1_tr1 = dict()\n",
    "tr1_tr2 = dict()\n",
    "\n",
    "tr2_tr2 = dict()\n",
    "tr2_tr1 = dict()\n",
    "\n",
    "tr1_prefix = \"ver28_GRCh37.\"\n",
    "tr2_prefix = \"ver38_GRCh38.\"\n",
    "\n",
    "def keepTheGreatest_local(current_gene, new_gene):\n",
    "    # print(f\"checking {current_gene} VS. {new_gene}\")\n",
    "    newGeneName, newGeneJacard, newGeneCont = new_gene\n",
    "    currentGeneName, currentGeneJacard, currentGeneCont = current_gene\n",
    "    newGeneScore = newGeneJacard + newGeneCont\n",
    "    currentGeneScore = currentGeneJacard + currentGeneCont\n",
    "    return new_gene if newGeneScore > currentGeneScore else current_gene\n",
    "\n",
    "def keepTheGreatest(current_list, new_gene):\n",
    "    current_list = current_list.copy()\n",
    "    newGeneName, newGeneJacard, newGeneCont = new_gene\n",
    "    newGeneScore = newGeneJacard + newGeneCont\n",
    "    maxScoreFound = float()\n",
    "    if not len(current_list):\n",
    "        return []\n",
    "\n",
    "    for currentGene in current_list:\n",
    "        currentGeneName, currentGeneJacard, currentGeneCont = currentGene\n",
    "        currentGeneScore = currentGeneJacard + currentGeneCont\n",
    "        maxScoreFound = max(currentGeneScore, maxScoreFound)\n",
    "    \n",
    "    if newGeneScore == maxScoreFound:\n",
    "        current_list.append(new_gene)\n",
    "    elif newGeneScore > maxScoreFound:\n",
    "        current_list = [new_gene]\n",
    "    \n",
    "    return current_list\n",
    "        \n",
    "        \n",
    "\n",
    "    \n",
    "\n",
    "file1 = \"sample.csv\"\n",
    "file2 = \"idx_merged_relations.csv\"\n",
    "\n",
    "with open(file2) as READER:\n",
    "    next(READER)\n",
    "    for line in READER:\n",
    "        line = line.split('|')\n",
    "        jaccard, containment = float(line[4]), float(line[5])\n",
    "        gene1, gene2 = line[1], line[6]\n",
    "        gene1_full = (gene1, jaccard, containment)\n",
    "        gene2_full = (gene2, jaccard, containment)\n",
    "\n",
    "\n",
    "        # gene1 & gene2 belongs to tr1\n",
    "        if tr1_prefix in gene1 and tr1_prefix in gene2:\n",
    "            if gene1 not in tr1_tr1:\n",
    "                tr1_tr1[gene1] = [gene2_full]\n",
    "            else:\n",
    "                tr1_tr1[gene1] = keepTheGreatest(tr1_tr1[gene1], gene2_full)\n",
    "        \n",
    "            if gene2 not in tr1_tr1:\n",
    "                tr1_tr1[gene2] = [gene1_full]\n",
    "            else:\n",
    "                tr1_tr1[gene2] = keepTheGreatest(tr1_tr1[gene2], gene1_full)\n",
    "\n",
    "                \n",
    "        \n",
    "        # gene1 & gene2 belongs to tr2\n",
    "        elif tr2_prefix in gene1 and tr2_prefix in gene2:\n",
    "            if gene1 not in tr2_tr2:\n",
    "                tr2_tr2[gene1] = [gene2_full]\n",
    "            else:\n",
    "                tr2_tr2[gene1] = keepTheGreatest(tr2_tr2[gene1], gene2_full)\n",
    "\n",
    "            if gene2 not in tr2_tr2:\n",
    "                tr2_tr2[gene2] = [gene1_full]\n",
    "            else:\n",
    "                tr2_tr2[gene2] = keepTheGreatest(tr2_tr2[gene2], gene1_full)\n",
    "        \n",
    "        elif tr2_prefix in gene1+gene2 and tr1_prefix in gene1+gene2:\n",
    "            if tr1_prefix in gene1:\n",
    "                if gene1 not in tr1_tr2:\n",
    "                    tr1_tr2[gene1] = [gene2_full]\n",
    "                else:\n",
    "                    tr1_tr2[gene1] = keepTheGreatest(tr1_tr2[gene1], gene2_full)\n",
    "\n",
    "                if gene2 not in tr2_tr1:\n",
    "                    tr2_tr1[gene2] = [gene1_full]\n",
    "                else:\n",
    "                    tr2_tr1[gene2] = keepTheGreatest(tr2_tr1[gene2], gene1_full)\n",
    "                \n",
    "            else:\n",
    "                if gene1 not in tr2_tr1:\n",
    "                    tr2_tr1[gene1] = [gene2_full]\n",
    "                else:\n",
    "                    tr2_tr1[gene1] = keepTheGreatest(tr2_tr1[gene1], gene2_full)\n",
    "\n",
    "                if gene2 not in tr1_tr2:\n",
    "                    tr1_tr2[gene2] = [gene1_full]\n",
    "                else:\n",
    "                    tr1_tr2[gene2] = keepTheGreatest(tr1_tr2[gene2], gene1_full)\n",
    "\n",
    "\n",
    "def getResults(tr_list, gene):\n",
    "    if gene in list(tr_list):\n",
    "        return tr_list[gene]\n",
    "    else:\n",
    "        return \"-\"\n",
    "\n",
    "\n",
    "_del = '\\t'\n",
    "\n",
    "with open(\"results_tr1.tsv\", 'w') as R:\n",
    "    R.write(f\"tr1{_del}tr1_local{_del}tr2_across\\n\")\n",
    "    for tr1_gene1, tr1_gene2 in tr1_tr1.items():\n",
    "        R.write(f\"{tr1_gene1}{_del}{tr1_gene2}{_del}{getResults(tr1_tr2, tr1_gene1)}\\n\")\n",
    "\n",
    "with open(\"results_tr2.tsv\", 'w') as R:\n",
    "    R.write(f\"tr2{_del}tr2_local{_del}tr1_across\\n\")\n",
    "    for tr2_gene1, tr2_gene2 in tr2_tr2.items():\n",
    "        R.write(f\"{tr2_gene1}{_del}{tr2_gene2}{_del}{getResults(tr2_tr1, tr2_gene1)}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17662f57-f9e9-49a0-b15d-7b7a24f9cb0b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
